jetcd-core/src/test/java/io/etcd/jetcd/impl/LeaseMemoryLeakTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2016-2021 The jetcd authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.etcd.jetcd.impl;
+
+import java.net.URI;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import io.etcd.jetcd.Client;
+import io.etcd.jetcd.Lease;
+import io.etcd.jetcd.test.EtcdClusterExtension;
+import io.restassured.RestAssured;
+
+import org.assertj.core.data.Percentage;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Timeout(value = 45, unit = TimeUnit.SECONDS)
+public class LeaseMemoryLeakTest {
+
+    @RegisterExtension
+    public final EtcdClusterExtension cluster = EtcdClusterExtension.builder()
+        .withNodes(1)
+        .build();
+
+    private Client client;
+    private Lease leaseClient;
+
+    private static final long TTL = 3;
+    private static final int ITERATIONS = 5;
+    private static final Pattern GO_ROUTINES_EXTRACT_PATTERN = Pattern.compile("go_goroutines (\\d+)");
+
+    @BeforeEach
+    public void setUp() {
+        client = TestUtil.client(cluster).build();
+        leaseClient = client.getLeaseClient();
+    }
+
+    @AfterEach
+    public void tearDown() {
+        if (client != null) {
+            client.close();
+        }
+    }
+
+    // https://github.com/etcd-io/jetcd/issues/1236
+    @Test
+    public void testKeepAliveOnceMemoryLeak() throws ExecutionException, InterruptedException {
+        final long leaseID = leaseClient.grant(TTL).get().getID();
+        final URI uri = cluster.cluster().clientEndpoints().get(0);
+
+        int before = extractGoRoutinesCount(uri);
+        assertThat(before).isGreaterThan(0);
+
+        for (int i = 0; i < ITERATIONS; i++) {
+            try {
+                leaseClient.keepAliveOnce(leaseID).get(1, TimeUnit.SECONDS);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        int after = extractGoRoutinesCount(uri);
+        assertThat(after).isGreaterThan(0);
+
+        assertThat(after).isGreaterThan(before);
```

Comment:
Not sure if this makes this a flake test as there will be some fluctuations here.